src/java/org/apache/cassandra/db/tries/SlicedTrie.java

```diff
@@ -0,0 +1,242 @@
+ * issuing the nodes and content from the source. As soon as the right bound becomes strictly smaller, we finish the
+ * walk.
+ *
+ * We don't explicitly construct tries for the two bounds; tracking the current depth (= prefix length) and transition
+ * as characters are requested from the key is sufficient as it is a trie with just a single descent path. Because we
+ * need the next character to tell if it's been exhausted, we keep these one position ahead. The source is always
+ * advanced, thus this gives us the thing to compare it against after the advance.
+ *
+ * We also track the current state to make some decisions a little simpler.
+ *
+ * See Trie.md for further details.
+ */
+public class SlicedTrie<T> extends Trie<T>
+{
+    private final Trie<T> source;
+
+    /** Left-side boundary. The characters of this are requested as we descend along the left-side boundary. */
+    private final ByteComparable left;
+
+    /** Right-side boundary. The characters of this are requested as we descend along the right-side boundary. */
+    private final ByteComparable right;
+
+    private final boolean includeLeft;
+    private final boolean includeRight;
+
+    public SlicedTrie(Trie<T> source, ByteComparable left, boolean includeLeft, ByteComparable right, boolean includeRight)
+    {
+        this.source = source;
+        this.left = left;
+        this.right = right;
+        this.includeLeft = includeLeft;
+        this.includeRight = includeRight;
+    }
+
+    @Override
+    protected Cursor<T> cursor()
+    {
+        return new SlicedCursor<>(this);
+    }
+
+    private enum State
+    {
+        /** The cursor is still positioned on some prefix of the left bound. Content should not be produced. */
+        ON_LEFT_BOUND,
+        /** The cursor is positioned inside the range, i.e. beyond the left bound, possibly on a prefix of the right. */
+        INSIDE,
+        /** The cursor is positioned beyond the right bound. Exhaustion (depth -1) has been reported. */
+        AFTER_RIGHT;
+    }
+
+    private static class SlicedCursor<T> implements Cursor<T>
+    {
+        private final ByteSource left;
+        private final ByteSource right;
+        private final boolean includeLeft;
+        private final boolean excludeRight;
+        private final Cursor<T> source;
+
+        State state;
+        int leftNext;
+        int leftDepth;
+        int rightNext;
+        int rightDepth;
+
+        public SlicedCursor(SlicedTrie<T> slicedTrie)
+        {
+            source = slicedTrie.source.cursor();
+            if (slicedTrie.left != null)
+            {
+                left = slicedTrie.left.asComparableBytes(BYTE_COMPARABLE_VERSION);
+                includeLeft = slicedTrie.includeLeft;
+                leftNext = left.next();
+                leftDepth = 1;
+                if (leftNext == ByteSource.END_OF_STREAM && includeLeft)
```

Comment:
They are different. `null` specifies the lack of bound on that side (and hence the boolean has no meaning in that case). Empty key is a valid key, and it is the very first in the ordering of keys. Because of that `("", inclusive)` as a left bound is exactly the same as `(null, anything)` but different from `("", exclusive)`.
On the right side `("", inclusive)` only permits the empty key, and `("", exclusive)` excludes everything.

Reply:
Thank you for explaining. It makes sense.